File: skills/init-knowledge-vault/scripts/knowledge_loom/audit.py
from __future__ import annotations

import re
import subprocess
from pathlib import Path

from .contract import load_note_frontmatter, validate_contract_data
from .focus import check_focus_view
from .models import Finding, Vault
from .pathing import (
    is_vault_relative_path,
    is_within_vault,
    resolve_vault_path,
    resolve_vault_pattern_prefix,
)


def _glob_regex(pattern: str) -> re.Pattern[str]:
    index = 0
    result = "^"
    while index < len(pattern):
        char = pattern[index]
        if char == "*":
            if index + 1 < len(pattern) and pattern[index + 1] == "*":
                index += 2
                if index < len(pattern) and pattern[index] == "/":
                    index += 1
                    result += "(?:.*/)?"
                else:
                    result += ".*"
                continue
            result += "[^/]*"
        elif char == "?":
            result += "[^/]"
        else:
            result += re.escape(char)
        index += 1
    return re.compile(result + "$")


def matches_path(pattern: str, path: str) -> bool:
    return bool(_glob_regex(pattern).match(path))


def _git(root: Path, *args: str) -> subprocess.CompletedProcess[str]:
    return subprocess.run(
        ["git", "-C", str(root), *args],
        text=True,
        capture_output=True,
        check=False,
    )


def _audit_declared_files(
    root: Path,
    values: object,
    *,
    boundary_code: str,
    boundary_message: str,
    missing_code: str,
    missing_message: str,
) -> list[Finding]:
    findings: list[Finding] = []
    if not isinstance(values, list):
        return findings
    for relative in values:
        if not isinstance(relative, str) or not is_vault_relative_path(relative):
            continue
        resolved = resolve_vault_path(root, relative)
        if resolved is None:
            findings.append(Finding("error", boundary_code, boundary_message, relative))
        elif not resolved.is_file():
            findings.append(Finding("error", missing_code, missing_message, relative))
    return findings


def audit_vault(vault: Vault) -> list[Finding]:
    findings = validate_contract_data(vault.contract)
    root = vault.root
    contract = vault.contract

    findings.extend(
        _audit_declared_files(
            root,
            contract.get("instruction_roots", []),
            boundary_code="path.instruction-boundary",
            boundary_message="instruction root resolves outside the vault",
            missing_code="path.instruction-root",
            missing_message="instruction root is missing",
        )
    )

    navigation = contract.get("navigation", {})
    if not isinstance(navigation, dict):
        navigation = {}
    findings.extend(
        _audit_declared_files(
            root,
            navigation.get("entrypoints", []),
            boundary_code="path.navigation-boundary",
            boundary_message="navigation entrypoint resolves outside the vault",
            missing_code="path.navigation",
            missing_message="navigation entrypoint is missing",
        )
    )

    subject_config = contract.get("subjects", {})
    if not isinstance(subject_config, dict):
        subject_config = {}
    subject_values = subject_config.get("values", [])
    subjects = set(subject_values) if isinstance(subject_values, list) else set()
    mode = subject_config.get("mode")
    checked: set[tuple[str, str]] = set()
    profiles = contract.get("metadata_profiles", {})
    if not isinstance(profiles, dict):
        profiles = {}
    for profile_name, profile in profiles.items():
        if not isinstance(profile, dict):
            continue
        severity = profile.get("severity", "error")
        required = profile.get("required", [])
        if not isinstance(required, list):
            required = []
        patterns = profile.get("paths", [])
        if not isinstance(patterns, list):
            patterns = []
        for pattern in patterns:
            if not isinstance(pattern, str) or not is_vault_relative_path(pattern):
                continue
            if resolve_vault_pattern_prefix(root, pattern) is None:
                findings.append(
                    Finding(
                        "error",
                        "path.metadata-boundary",
                        f"profile `{profile_name}` path prefix resolves outside the vault",
                        pattern,
                    )
                )
                continue
            try:
                matches = root.glob(pattern)
                paths = list(matches)
            except (OSError, RuntimeError, ValueError) as exc:
                findings.append(
                    Finding(
                        "error",
                        "metadata.pattern",
                        f"profile `{profile_name}` has an unusable path pattern: {exc}",
                        pattern,
                    )
                )
                continue
            for path in paths:
                if not is_within_vault(root, path):
                    findings.append(
                        Finding(
                            "error",
                            "path.metadata-boundary",
                            f"profile `{profile_name}` matched a file outside the vault",
                            path.relative_to(root).as_posix(),
                        )
                    )
                    continue
                if not path.is_file() or path.suffix.casefold() != ".md":
                    continue
                relative = path.relative_to(root).as_posix()
                key = (profile_name, relative)
                if key in checked:
                    continue
                checked.add(key)
                metadata = load_note_frontmatter(path)
                for field in required:
                    if field not in metadata:
                        findings.append(
                            Finding(
                                severity,
                                "metadata.missing",
                                f"profile `{profile_name}` requires `{field}`",
                                relative,
                            )
                        )
                subject_key = "owner" if "owner" in metadata else "subject" if "subject" in metadata else None
                if mode == "multiple" and subject_key and metadata[subject_key] not in subjects:
                    findings.append(
                        Finding(
                            "error",
                            "metadata.subject",
                            f"`{subject_key}` is not declared in contract subjects",
                            relative,
                        )
                    )

    privacy = contract.get("privacy", {})
    never_track = privacy.get("never_track", []) if isinstance(privacy, dict) else []
    if not isinstance(never_track, list):
        never_track = []
    valid_never_track: list[str] = []
    for pattern in never_track:
        if not isinstance(pattern, str) or not is_vault_relative_path(pattern):
            continue
        if resolve_vault_pattern_prefix(root, pattern) is None:
            findings.append(
                Finding(
                    "error",
                    "path.privacy-boundary",
                    "privacy path prefix resolves outside the vault",
                    pattern,
                )
            )
            continue
        valid_never_track.append(pattern)

    history = contract.get("history", {})
    history_type = history.get("type") if isinstance(history, dict) else None
    git_check = _git(root, "rev-parse", "--show-toplevel")
    is_git = (
        git_check.returncode == 0
        and bool(git_check.stdout.strip())
        and Path(git_check.stdout.strip()).resolve() == root.resolve()
    )
    if history_type == "git" and not is_git:
        findings.append(Finding("error", "git.missing", "contract requires Git but root is not a Git repository"))
    if history_type == "none" and is_git:
        findings.append(Finding("warning", "git.unconfigured", "root is a Git repository but contract declares no history"))

    if is_git:
        status = _git(root, "status", "--short")
        if status.stdout.strip():
            findings.append(Finding("info", "git.dirty", "working tree has uncommitted changes"))
        tracked = _git(root, "ls-files")
        for relative in tracked.stdout.splitlines():
            for pattern in valid_never_track:
                if matches_path(pattern, relative):
                    findings.append(
                        Finding(
                            "error",
                            "privacy.tracked",
                            f"tracked path matches privacy rule `{pattern}`",
                            relative,
                        )
                    )

    focus_views = contract.get("focus_views", {})
    if not isinstance(focus_views, dict):
        focus_views = {}
    for name, view in focus_views.items():
        if not isinstance(view, dict) or not isinstance(view.get("path"), str):
            continue
        if not is_vault_relative_path(view["path"]):
            continue
        findings.extend(check_focus_view(root, name, view))

    return findings

File: src/knowledge-loom/audit.ts
import fs from "node:fs";
import path from "node:path";

import { finding, isUnknownRecord, loadNoteFrontmatter, validateContractData } from "./contract.js";
import { runDeclaredContentCheck } from "./content-checks.js";
import { errorMessage } from "./errors.js";
import { checkFocusView, parseFocusView } from "./focus.js";
import { gitOutput } from "./git.js";
import {
  canonicalPath,
  isVaultRelativePath,
  isWithin,
  resolveVaultPath,
  resolveVaultPatternPrefix,
  toPosixRelative,
} from "./pathing.js";
import type { Finding, FindingSeverity, LoadedVault } from "./types.js";

function escapeRegex(character: string): string {
  return /[\\^$.*+?()[\]{}|]/.test(character) ? `\\${character}` : character;
}

function globRegex(pattern: string, { characterClasses = true }: { characterClasses?: boolean } = {}): RegExp {
  let index = 0;
  let result = "^";
  while (index < pattern.length) {
    const character = pattern[index]!;
    if (character === "*") {
      if (pattern[index + 1] === "*") {
        index += 2;
        if (pattern[index] === "/") {
          index += 1;
          result += "(?:.*/)?";
        } else {
          result += ".*";
        }
        continue;
      }
      result += "[^/]*";
    } else if (character === "?") {
      result += "[^/]";
    } else if (character === "[" && characterClasses) {
      let close = index + 1;
      if (["!", "^"].includes(pattern[close] ?? "")) close += 1;
      if (pattern[close] === "]") close += 1;
      close = pattern.indexOf("]", close);
      if (close < 0) {
        result += "\\[";
      } else {
        let contents = pattern.slice(index + 1, close);
        let prefix = "";
        if (contents.startsWith("!")) {
          prefix = "^";
          contents = contents.slice(1);
        } else if (contents.startsWith("^")) {
          contents = `\\${contents}`;
        }
        let escapedContents = contents.replaceAll("\\", "\\\\").replaceAll("[", "\\[");
        if (escapedContents.startsWith("]")) escapedContents = `\\${escapedContents}`;
        result += `[${prefix}${escapedContents}]`;
        index = close;
      }
    } else {
      result += escapeRegex(character);
    }
    index += 1;
  }
  try {
    return new RegExp(`${result}$`);
  } catch (error) {
    if (!characterClasses) throw error;
    return globRegex(pattern, { characterClasses: false });
  }
}

export function matchesPath(pattern: string, candidate: string): boolean {
  return globRegex(pattern).test(candidate);
}

interface DeclaredFilesOptions {
  boundaryCode: string;
  boundaryMessage: string;
  missingCode: string;
  missingMessage: string;
}

function auditDeclaredFiles(root: string, values: unknown, {
  boundaryCode,
  boundaryMessage,
  missingCode,
  missingMessage,
}: DeclaredFilesOptions): Finding[] {
  if (!Array.isArray(values)) return [];
  const findings: Finding[] = [];
  for (const relative of values) {
    if (typeof relative !== "string" || !isVaultRelativePath(relative)) continue;
    const resolved = resolveVaultPath(root, relative);
    if (resolved === null) findings.push(finding("error", boundaryCode, boundaryMessage, relative));
    else if (!fs.statSync(resolved, { throwIfNoEntry: false })?.isFile()) findings.push(finding("error", missingCode, missingMessage, relative));
  }
  return findings;
}

function patternPrefix(patternValue: string): string[] {
  const parts: string[] = [];
  for (const part of patternValue.split("/")) {
    if (/[*?[\]]/.test(part)) break;
    parts.push(part);
  }
  return parts;
}

function walk(root: string): string[] {
  const result: string[] = [];
  if (!fs.existsSync(root)) return result;
  const visit = (current: string): void => {
    for (const entry of fs.readdirSync(current, { withFileTypes: true })) {
      const candidate = path.join(current, entry.name);
      result.push(candidate);
      if (entry.isDirectory() && !entry.isSymbolicLink()) visit(candidate);
    }
  };
  if (fs.lstatSync(root).isDirectory() && !fs.lstatSync(root).isSymbolicLink()) visit(root);
  else result.push(root);
  return result;
}

function globPaths(root: string, patternValue: string): string[] {
  if (!/[*?[]/.test(patternValue)) {
    const candidate = path.join(root, ...patternValue.split("/"));
    return fs.existsSync(candidate) ? [candidate] : [];
  }
  const prefix = patternPrefix(patternValue);
  const searchRoot = path.join(root, ...prefix);
  return walk(searchRoot).filter((candidate) => matchesPath(patternValue, toPosixRelative(root, candidate)));
}

export async function auditVault(
  vault: LoadedVault,
  { registryPath }: { registryPath?: string | undefined } = {},
): Promise<Finding[]> {
  const findings = validateContractData(vault.contract);
  const { root, contract } = vault;

  findings.push(...auditDeclaredFiles(root, contract.instruction_roots ?? [], {
    boundaryCode: "path.instruction-boundary",
    boundaryMessage: "instruction root resolves outside the vault",
    missingCode: "path.instruction-root",
    missingMessage: "instruction root is missing",
  }));

  const navigation = isUnknownRecord(contract.navigation) ? contract.navigation : {};
  findings.push(...auditDeclaredFiles(root, navigation.entrypoints ?? [], {
    boundaryCode: "path.navigation-boundary",
    boundaryMessage: "navigation entrypoint resolves outside the vault",
    missingCode: "path.navigation",
    missingMessage: "navigation entrypoint is missing",
  }));

  const subjectConfig = isUnknownRecord(contract.subjects) ? contract.subjects : {};
  const subjectValues = Array.isArray(subjectConfig.values)
    ? subjectConfig.values.filter((value): value is string => typeof value === "string")
    : [];
  const subjects = new Set(subjectValues);
  const profiles = isUnknownRecord(contract.metadata_profiles) ? contract.metadata_profiles : {};
  const checked = new Set<string>();
  for (const [profileName, profile] of Object.entries(profiles)) {
    if (!isUnknownRecord(profile)) continue;
    const severity: FindingSeverity = profile.severity === "warning" ? "warning" : "error";
    const required = Array.isArray(profile.required)
      ? profile.required.filter((value): value is string => typeof value === "string")
      : [];
    const patterns = Array.isArray(profile.paths) ? profile.paths : [];
    for (const patternValue of patterns) {
      if (typeof patternValue !== "string" || !isVaultRelativePath(patternValue)) continue;
      if (resolveVaultPatternPrefix(root, patternValue) === null) {
        findings.push(finding("error", "path.metadata-boundary", `profile \`${profileName}\` path prefix resolves outside the vault`, patternValue));
        continue;
      }
      let paths;
      try {
        paths = globPaths(root, patternValue);
      } catch (error) {
        findings.push(finding("error", "metadata.pattern", `profile \`${profileName}\` has an unusable path pattern: ${errorMessage(error)}`, patternValue));
        continue;
      }
      for (const candidate of paths) {
        const relative = toPosixRelative(root, candidate);
        if (!isWithin(root, candidate)) {
          findings.push(finding("error", "path.metadata-boundary", `profile \`${profileName}\` matched a file outside the vault`, relative));
          continue;
        }
        if (!fs.statSync(candidate, { throwIfNoEntry: false })?.isFile() || path.extname(candidate).toLocaleLowerCase() !== ".md") continue;
        const key = `${profileName}\0${relative}`;
        if (checked.has(key)) continue;
        checked.add(key);
        const metadata = loadNoteFrontmatter(candidate);
        for (const field of required) {
          if (!Object.hasOwn(metadata, field)) findings.push(finding(severity, "metadata.missing", `profile \`${profileName}\` requires \`${field}\``, relative));
        }
        const subjectKey = Object.hasOwn(metadata, "owner") ? "owner" : Object.hasOwn(metadata, "subject") ? "subject" : null;
        const metadataSubject = subjectKey ? metadata[subjectKey] : undefined;
        if (subjectConfig.mode === "multiple" && subjectKey && (typeof metadataSubject !== "string" || !subjects.has(metadataSubject))) {
          findings.push(finding("error", "metadata.subject", `\`${subjectKey}\` is not declared in contract subjects`, relative));
        }
      }
    }
  }

  const privacy = isUnknownRecord(contract.privacy) ? contract.privacy : {};
  const neverTrack = Array.isArray(privacy.never_track) ? privacy.never_track : [];
  const validNeverTrack: string[] = [];
  for (const patternValue of neverTrack) {
    if (typeof patternValue !== "string" || !isVaultRelativePath(patternValue)) continue;
    if (resolveVaultPatternPrefix(root, patternValue) === null) {
      findings.push(finding("error", "path.privacy-boundary", "privacy path prefix resolves outside the vault", patternValue));
      continue;
    }
    validNeverTrack.push(patternValue);
  }

  const history = isUnknownRecord(contract.history) ? contract.history : {};
  const gitRoot = gitOutput(root, ["rev-parse", "--show-toplevel"])?.trim() ?? "";
  const isGit = Boolean(gitRoot) && canonicalPath(gitRoot) === canonicalPath(root);
  if (history.type === "git" && !isGit) findings.push(finding("error", "git.missing", "contract requires Git but root is not a Git repository"));
  if (history.type === "none" && isGit) findings.push(finding("warning", "git.unconfigured", "root is a Git repository but contract declares no history"));

  if (isGit) {
    const statusOutput = gitOutput(root, ["status", "--short"]);
    if (statusOutput?.trim()) findings.push(finding("info", "git.dirty", "working tree has uncommitted changes"));
    const trackedFilesOutput = gitOutput(root, ["ls-files"]);
    for (const relative of (trackedFilesOutput ?? "").split(/\r?\n/).filter(Boolean)) {
      for (const patternValue of validNeverTrack) {
        if (matchesPath(patternValue, relative)) findings.push(finding("error", "privacy.tracked", `tracked path matches privacy rule \`${patternValue}\``, relative));
      }
    }
  }

  const focusViews = isUnknownRecord(contract.focus_views) ? contract.focus_views : {};
  for (const [name, value] of Object.entries(focusViews)) {
    const view = parseFocusView(value);
    if (!view || !isVaultRelativePath(view.path)) continue;
    findings.push(...checkFocusView(root, name, view));
  }
  if (!findings.some((item) => item.severity === "error")) {
    findings.push(...await runDeclaredContentCheck(vault, { registryPath }));
  }
  return findings;
}
